xarray/core/parallel.py

```diff
@@ -0,0 +1,275 @@
+try:
+    import dask
+    import dask.array
+    from dask.highlevelgraph import HighLevelGraph
+
+except ImportError:
+    pass
+
+import itertools
+import numpy as np
+import operator
+
+from .dataarray import DataArray
+from .dataset import Dataset
+
+
+def _to_array(obj):
+    if not isinstance(obj, Dataset):
+        raise ValueError("Trying to convert DataArray to DataArray!")
+
+    if len(obj.data_vars) > 1:
+        raise ValueError(
+            "Trying to convert Dataset with more than one variable to DataArray"
+        )
+
+    name = list(obj.data_vars)[0]
+    # this should be easier
+    da = obj.to_array().squeeze().drop("variable")
+    da.name = name
+    return da
+
+
+def make_meta(obj):
+
+    from dask.array.utils import meta_from_array
+
+    if isinstance(obj, DataArray):
+        meta = DataArray(obj.data._meta, dims=obj.dims)
+
+    if isinstance(obj, Dataset):
+        meta = Dataset()
+        for name, variable in obj.variables.items():
+            if dask.is_dask_collection(variable):
+                meta_obj = obj[name].data._meta
+            else:
+                meta_obj = meta_from_array(obj[name].data)
+            meta[name] = DataArray(meta_obj, dims=obj[name].dims)
+    else:
+        meta = obj
+
+    return meta
+
+
+def infer_template(func, obj, *args, **kwargs):
+    """ Infer return object by running the function on meta objects. """
+    meta_args = []
+    for arg in (obj,) + args:
+        meta_args.append(make_meta(arg))
+
+    try:
+        template = func(*meta_args, **kwargs)
+    except ValueError:
+        raise ValueError("Cannot infer object returned by user-provided function.")
+
+    return template
+
+
+def _make_dict(x):
+    # Dataset.to_dict() is too complicated
+    # maps variable name to numpy array
+    if isinstance(x, DataArray):
+        x = x._to_temp_dataset()
+
+    to_return = dict()
+    for var in x.variables:
+        # if var not in x:
+        #    raise ValueError("Variable %r not found in returned object." % var)
+        to_return[var] = x[var].values
+
+    return to_return
+
+
+def map_blocks(func, obj, *args, **kwargs):
+    """
+    Apply a function to each chunk of a DataArray or Dataset. This function is experimental
+    and its signature may change.
+
+    Parameters
+    ----------
+    func: callable
+        User-provided function that should accept DataArrays corresponding to one chunk.
+        The function will be run on a small piece of data that looks like 'obj' to determine
+        properties of the returned object such as dtype, variable names,
+        new dimensions and new indexes (if any).
+
+        This function cannot
+        - change size of existing dimensions.
+        - add new chunked dimensions.
+
+    obj: DataArray, Dataset
+        Chunks of this object will be provided to 'func'. The function must not change
+        shape of the provided DataArray.
+    args:
+        Passed on to func.
```

Comment:
I would suggest suggest dedicated keyword arguments for `args` and `kwargs` rather than using `*args, **kwargs`. The problem `*args, **kwargs` is that it limits our ability to add new arguments later, because we've already promised we're passing on everything.